src/GafferScene/SceneWriter.cpp

```diff
@@ -53,85 +56,101 @@ using namespace GafferScene;
+	/// once all the data has been read then we take a global lock and write
+	/// into the SceneInterface
+	bool operator()( const ScenePlug *scene, const ScenePlug::ScenePath &scenePath )
 	{
-		ConstCompoundObjectPtr globals = scene->globalsPlug()->getValue();
-		output->writeAttribute( "gaffer:globals", globals.get(), time );
-	}
+		ContextPtr context = new Context( *Context::current() );
+		Context::Scope scopedContext( context.get() );
 
-	ConstObjectPtr object = scene->objectPlug()->getValue();
+		ConstCompoundObjectPtr attributes = scene->attributesPlug()->getValue();
 
-	if( object->typeId() != IECore::NullObjectTypeId && scenePath.size() > 0 )
-	{
-		output->writeObject( object.get(), time );
-	}
+		ConstCompoundObjectPtr globals;
 
-	Imath::Box3f b = scene->boundPlug()->getValue();
+		ConstObjectPtr object = scene->objectPlug()->getValue();
+		Imath::Box3f bound = scene->boundPlug()->getValue();
 
-	output->writeBound( Imath::Box3d( Imath::V3f( b.min ), Imath::V3f( b.max ) ), time );
+		IECore::M44dDataPtr transformData;
 
-	if( scenePath.size() )
-	{
-		Imath::M44f t = scene->transformPlug()->getValue();
-		Imath::M44d transform(
-			t[0][0], t[0][1], t[0][2], t[0][3],
-			t[1][0], t[1][1], t[1][2], t[1][3],
-			t[2][0], t[2][1], t[2][2], t[2][3],
-			t[3][0], t[3][1], t[3][2], t[3][3]
-		);
-
-		output->writeTransform( new IECore::M44dData( transform ), time );
-	}
+		if( scenePath.empty() )
+		{
+			globals = scene->globalsPlug()->getValue();
+		}
+		else
+		{
+			Imath::M44f t = scene->transformPlug()->getValue();
+			transformData = new IECore::M44dData( Imath::M44d (
+				t[0][0], t[0][1], t[0][2], t[0][3],
+				t[1][0], t[1][1], t[1][2], t[1][3],
+				t[2][0], t[2][1], t[2][2], t[2][3],
+				t[3][0], t[3][1], t[3][2], t[3][3]
+			) );
+		}
 
-	// query the set membership for this location and write as tags.
-	const std::vector<InternedString> &setNames = scene->setNamesPlug()->getValue()->readable();
+		// query the set membership for this location and write as tags.
+		const std::vector<InternedString> &setNames = scene->setNamesPlug()->getValue()->readable();
 
-	SceneInterface::NameList locationSets;
-	locationSets.reserve( setNames.size() );
+		SceneInterface::NameList locationSets;
+		locationSets.reserve( setNames.size() );
 
-	for( size_t i = 0, numSets = setNames.size(); i < numSets; ++i )
-	{
-		context->set( ScenePlug::setNameContextName, setNames[i] );
-		const PathMatcher &pathMatcher = scene->setPlug()->getValue()->readable();
+		for( size_t i = 0, numSets = setNames.size(); i < numSets; ++i )
+		{
+			context->set( ScenePlug::setNameContextName, setNames[i] );
+			const PathMatcher &pathMatcher = scene->setPlug()->getValue()->readable();
+
+			if( pathMatcher.match( scenePath ) == Filter::ExactMatch )
+			{
+				locationSets.push_back( setNames[i] );
+			}
+		}
 
-		if( pathMatcher.match( scenePath ) == Filter::ExactMatch )
+		tbb::mutex::scoped_lock l( g_writerMutex );
+
+		if ( !scenePath.empty() )
```

Comment:
Since you once nitpicked my whitespacing, I'll take the same liberty :) The prevailing style in Gaffer is : 

```
if( !scenePath.empty() )
{
}
```